python/pyfury/meta/metastring.py

```diff
@@ -0,0 +1,324 @@
+        return decoded_str.capitalize()
+
+    def _decode_rep_all_to_lower_special(self, data):
+        decoded_str = self._decode_lower_special(data)
+        result = []
+        skip = False
+        for i, char in enumerate(decoded_str):
+            if skip:
+                skip = False
+                continue
+            if char == "|":
+                result.append(decoded_str[i + 1].upper())
+                skip = True
+            else:
+                result.append(char)
+        return "".join(result)
+
+
+# Encodes plain text strings into MetaString objects with specified encoding mechanisms.
+class MetaStringEncoder:
+    def encode(self, input_string):
+
+        # Long meta string than 32767 is not allowed.
+        assert (
+            len(input_string) < _METASTRING_NUM_CHARS_LIMIT
+        ), "Long meta string than 32767 is not allowed."
+
+        if not input_string:
+            return MetaString(input_string, Encoding.UTF_8, np.array([], np.uint8), 0)
+
+        encoding = self.compute_encoding(input_string)
+        return self.encode_with_encoding(input_string, encoding)
+
+    def encode_with_encoding(self, input_string, encoding):
+
+        # Long meta string than 32767 is not allowed.
+        assert (
+            len(input_string) < _METASTRING_NUM_CHARS_LIMIT
+        ), "Long meta string than 32767 is not allowed."
+
+        if not input_string:
+            return MetaString(input_string, Encoding.UTF_8, np.array([], np.uint8), 0)
+
+        length = len(input_string)
+        if encoding == Encoding.LOWER_SPECIAL:
+            encoded_data = self._encode_lower_special(input_string)
+            return MetaString(input_string, encoding, encoded_data, length * 5)
+        elif encoding == Encoding.LOWER_UPPER_DIGIT_SPECIAL:
+            encoded_data = self._encode_lower_upper_digit_special(input_string)
+            return MetaString(input_string, encoding, encoded_data, length * 6)
+        elif encoding == Encoding.FIRST_TO_LOWER_SPECIAL:
+            encoded_data = self._encode_first_to_lower_special(input_string)
+            return MetaString(input_string, encoding, encoded_data, length * 5)
+        elif encoding == Encoding.ALL_TO_LOWER_SPECIAL:
+            chars = list(input_string)
+            upper_count = sum(1 for c in chars if c.isupper())
+            encoded_data = self._encode_all_to_lower_special(chars, upper_count)
+            return MetaString(
+                input_string, encoding, encoded_data, (upper_count + length) * 5
+            )
+        else:
+            encoded_data = np.frombuffer(input_string.encode("utf-8"), dtype=np.uint8)
+            return MetaString(
+                input_string, Encoding.UTF_8, encoded_data, len(encoded_data) * 8
+            )
+
+    def compute_encoding(self, input_string):
+        if not input_string:
+            return Encoding.LOWER_SPECIAL
+
+        chars = list(input_string)
+        statistics = self._compute_statistics(chars)
```

Comment:
Could you return a named tuple instead of a dict. And we'd better use encoding enum instead of string